test/gtest/ucp/test_ucp_am.cc

```diff
@@ -799,6 +800,192 @@ UCS_TEST_P(test_ucp_am_nbx, rx_am_mpools,
+        request_wait(pending_sptr);
+        pending_sptr = NULL;
+
+        ucs_time_t deadline = ucs_get_time() +
+                              ucs_time_from_sec(timeout) * ucs::test_time_multiplier();
+        while ((ucs_get_time() < deadline) && pending_sptr == send_completed) {
+            pending_sptr = send_am(sdt_desc, get_send_flag() | flags,
+                                            m_hdr_copy.data(), header_size, memh);
+        }
+
+        return pending_sptr;
+    }    
+protected:
+    void check_eager(size_t size, size_t header_size = 0ul,
+                           unsigned flags = 0, unsigned data_cb_flags = 0) {
+        mem_buffer sbuf(size, tx_memtype());
+        sbuf.pattern_fill(SEED);
+        m_hdr_copy.resize(header_size);
+        ucs::fill_random(m_hdr_copy);
+        m_hdr = m_hdr_copy;
+        ucs_status_ptr_t pending_sptr = NULL;
+        req_counter = 0;
+        cb_counter = 0;
+
+        ucp::data_type_desc_t sdt_desc(m_dt, sbuf.ptr(), size);                               
+        ucp_mem_h memh;
+
+        if (prereg()) {
+            memh = sender().mem_map(sbuf.ptr(), size);
+        }
+
+        set_am_data_handler(receiver(), TEST_AM_NBX_ID, am_data_cb, this,
+                                    data_cb_flags);
+
+        pending_sptr = fill_sq(sdt_desc, memh, header_size, get_send_flag() | flags, data_cb_flags);
+
+        if (pending_sptr == send_completed) {
+            UCS_TEST_MESSAGE <<"Failed to get pending request";
+        }
+
+        ucs::fill_random(m_hdr_copy);
+
+        while(progress());
+        wait_for_value(&cb_counter, req_counter);
+        request_wait(pending_sptr);
+        EXPECT_EQ(req_counter, cb_counter);
+
+        if (prereg()) {
+            sender().mem_unmap(memh);
+        }
+    }
+
+    void check_rndv(size_t size, size_t header_size = 0ul,
+                           unsigned flags = 0, unsigned data_cb_flags = 0) {
+        mem_buffer sbuf(size, tx_memtype());
+        sbuf.pattern_fill(SEED);
+        m_hdr_copy.resize(header_size*2);
+        ucs::fill_random(m_hdr_copy);
+        m_hdr = m_hdr_copy;
+        ucs_status_ptr_t rndv_pending_sptr = NULL;
+        ucs_status_ptr_t pending_sptr = NULL;
+        req_counter = 0;
+        ucp::data_type_desc_t sdt_desc(m_dt, sbuf.ptr(), size);                               
+        ucp_mem_h memh;
+
+        if (prereg()) {
+            memh = sender().mem_map(sbuf.ptr(), size);
+        }
+
+        set_am_data_handler(receiver(), TEST_AM_NBX_ID, am_data_cb, this,
+                                    data_cb_flags);
+
+        pending_sptr = fill_sq(sdt_desc, memh, header_size, get_send_flag() | flags, data_cb_flags);
+
+        if (pending_sptr == send_completed) {
+            UCS_TEST_MESSAGE <<"Failed to get pending request";
+        }
+
+        rndv_pending_sptr = send_am(sdt_desc, get_send_flag() | flags,
+                                            m_hdr_copy.data(), header_size*2, memh);
```

Comment:
`header_size*2` is confusing.
When one calls `check_rndv` with a particular `header_size`, it is not supposed to "hide" x2 inside.